### ISSUE TITLE
ERROR from conch test when pycrypto is not installed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3568]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)| @zooko reported|
|-|-|
|Trac ID|trac#3568|
|Type|defect|
|Created|2008-12-09 12:56:15Z|
|Branch|https://github.com/twisted/twisted/tree/optional-crypto-3568-2|

Apparently the beneficial effect of [24770] has partly regressed.  Here is a patch which causes those tests to be skipped when pycrypto cannot be imported.

Attachments:

* [diff.patch.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/717/717d8d71766b2dc8682457e7cf5d40db0b0cb08a/3a46c6867628d521789f91ef77c812812cd69c81.txt) (74612 bytes) - added by zooko on 2008-12-09 12:56:44Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3568 3568
type__defect defect
reporter__zooko zooko
priority__normal normal
milestone__ 
branch__branches_optional_crypto_3568_2 branches/optional-crypto-3568-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1228827375000000 1228827375000000
changetime__1233106970000000 1233106970000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

No builder configuration is testing this.  That's how it was able to regress.  Fixing it over and over again is kind of a waste of time.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_2"></a>|@zooko commented|
|-|-|

Well, I guess I could contribute a "definitely doesn't have pycrypto" installed buildslave, and then remember not to go ahead and install pycrypto on it later...  Perhaps it would be better to write a builder that creates a dummy Crypto module thta raises ImportError.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @exarkun|
|-|-|

I can probably set up such a builder on one of the slaves we have.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_4"></a>|@zooko commented|
|-|-|

Great!  And by the way, pycrypto has bugs such that it sometimes gives incorrect output from its crypto algorithms.  Be alert!  The world needs more lerts.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Yea. :/  I recently learned that we're misusing the PyCrypto randomness APIs and we need to stop.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Slave is set up, http://buildbot.twistedmatrix.com/builders/py2.5-without-modules

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [25637]) Branching to 'optional-crypto-3568'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [25638]) Apply the aptly named diff.patch.txt

refs [#3568](https://github.com/twisted/twisted/issues/3568)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [26127]) Branching to 'optional-crypto-3568-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

A somewhat shorter solution than the previous proposal is in the -2 branch.  Build results are at <http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/optional-crypto-3568-2>, plus a build without PyCrypto available had these results: <http://buildbot.twistedmatrix.com/builders/py2.5-without-modules/builds/49>.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj set owner to @exarkun|
|-|-|

Looks good, I think you can go ahead and merge this.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

(In [26132]) Merge optional-crypto-3568-2

Author: exarkun
Reviewer: jonothanj
Fixes: [#3568](https://github.com/twisted/twisted/issues/3568)

Change two recently added Conch test modules to handle missing dependencies
more gracefully (by skipping instead of failing).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
